src/storage/table/scan_state.cpp

```diff
@@ -174,11 +175,108 @@ ParallelCollectionScanState::ParallelCollectionScanState()
+	return row_groups->GetRootSegment();
+}
+
+RowGroup *ParallelCollectionScanState::GetNextRowGroup(const shared_ptr<RowGroupSegmentTree> &row_groups,
+                                                       RowGroup *row_group) {
+	return row_groups->GetNextSegment(row_group);
+}
+
+CustomOrderParallelCollectionScanState::CustomOrderParallelCollectionScanState(const RowGroupOrderOptions &options)
+    : column_idx(options.column_idx), order_by(options.order_by), order_type(options.order_type),
+      column_type(options.column_type), root(nullptr) {
+}
+
+RowGroup *CustomOrderParallelCollectionScanState::GetRootSegment(const shared_ptr<RowGroupSegmentTree> &row_groups) {
+	if (!root) {
+		D_ASSERT(row_groups);
+		auto preordered_row_groups =
+		    CollectionScanStateHelper::OrderRowGroups(row_groups.get(), column_idx, order_by, column_type);
+		ordered_row_groups = CollectionScanStateHelper::CreateRowGroupMap(preordered_row_groups, order_type, root);
+	}
+	return root;
+}
+
+RowGroup *CustomOrderParallelCollectionScanState::GetNextRowGroup(const shared_ptr<RowGroupSegmentTree> &row_groups,
+                                                                  RowGroup *row_group) {
+	return ordered_row_groups[row_group];
+}
+
+Value CollectionScanStateHelper::RetrieveStat(const BaseStatistics &stats, OrderByStatistics order_by,
+                                              OrderByColumnType column_type) {
+	switch (order_by) {
+	case OrderByStatistics::MIN:
+		return column_type == OrderByColumnType::NUMERIC ? NumericStats::Min(stats) : StringStats::Min(stats);
+	case OrderByStatistics::MAX:
+		return column_type == OrderByColumnType::NUMERIC ? NumericStats::Max(stats) : StringStats::Max(stats);
+	}
+	return Value();
+}
+
+std::multimap<Value, RowGroup *> CollectionScanStateHelper::OrderRowGroups(RowGroupSegmentTree *row_groups,
+                                                                           column_t column_idx,
+                                                                           OrderByStatistics order_by,
+                                                                           OrderByColumnType column_type) {
+	std::multimap<Value, RowGroup *> ordered_row_groups;
+	auto row_group = row_groups->GetRootSegment();
+	while (row_group) {
+		auto stats = row_group->GetStatistics(column_idx);
+		Value comparison_value = RetrieveStat(*stats, order_by, column_type);
+		ordered_row_groups.emplace(comparison_value, row_group);
+		row_group = row_groups->GetNextSegment(row_group);
+	}
+
+	return ordered_row_groups;
+}
+
+unordered_map<RowGroup *, RowGroup *>
+CollectionScanStateHelper::CreateRowGroupMap(const multimap<Value, RowGroup *> &row_groups,
+                                             RowGroupOrderType order_type, RowGroup *&root) {
+	if (row_groups.empty()) {
+		return {};
+	}
+
+	unordered_map<RowGroup *, RowGroup *> row_group_map;
```

Comment:
Instead of doing this remap, since we have already obtained a (sorted) list of row groups, it would be better to fetch row groups from that list and iterate over that list in-order.

Relying on a re-mapping might be problematic, since we are traversing the row group list multiple times, and the second time might give us different row groups. I'm not sure how this would work out if e.g. another connection appends a new row group to the table while we are scanning it. We should probably add a test for this at the very least.